src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,450 @@
+            rdoc['rses'] = rses
+
+        # for each rule matching the DID + Rucio account
+        for rdoc in rules:
+            rid = rdoc['id']
+            rses = rdoc['rses']
+            for rse in rses:
+                # if rule RSE is not in expectedRSEs, then this rule needs to be deleted
+                if rse not in expectedRSEs:
+                    # upon successful rule deletion, also remove the RSE name from currentRSEs
+                    if isinstance(rucioClient, Client):
+                        rucioClient.delete_replication_rule(rid)
+                    else:
+                        rucioClient.deleteRule(rid)
+                    if rse in doc['currentRSEs']:
+                        doc['currentRSEs'].remove(rse)
+                    msg = f"rse {rse} rule is deleted and remove from currentRSEs"
+                    logger.info(msg)
+                    # delete rid in ruleIds (if any)
+                    if rid in doc['ruleIds']:
+                        doc['ruleIds'].remove(rid)
+                        msg = f"rule Id {rid} is removed from ruleIds"
+                        logger.info(msg)
+                else:
+                    # else, save this rule RSE in the local copy of currentRSEs.
+                    # This rule is likely still being processed by Rucio.
+                    if rse not in doc['currentRSEs']:
+                        doc['currentRSEs'].append(rse)
+                        msg = f"rse {rse} is added to currentRSEs"
+                        logger.info(msg)
+
+        # for each expectedRSEs not in localCopyOfCurrentRSEs
+        for rse in set(expectedRSEs).difference(localCopyOfCurrentRSEs):
+            # check whether the RSE has enough available space available for that
+            if isinstance(rucioClient, Client):
+                gen = rucioClient.get_rse_usage(rse)
+            else:
+                gen = rucioClient.getRSEUsage(rse)
+            enoughSpace = False
+            for rec in gen:
+                # we'll use total - used space to exceed pileupSize with some margin
+                if rec['total'] - rec['used'] - pileupSize > marginSpace:
+                    enoughSpace = True
+                    break
+            dids = [inputArgs]
+            if enoughSpace:
+                msg = f"active task {uuod}, for dids {dids} there is enough space at RSE {rse}"
+                logger.warning(msg)
+                # create the rule and append the rule id to ruleIds
+                kwargs = rucioArgs(rucioAccount)
+                if isinstance(rucioClient, Client):
+                    copies = 1
+                    rids = rucioClient.add_replication_rule(dids, copies, rse, **kwargs)
+                else:
+                    rids = rucioClient.createReplicationRule(name, rse, **kwargs)
+                # add new ruleId to document ruleIds
+                for rid in rids:
```

Comment:
maybe around these lines we should set `modify=True`?

Reply:
corrected